builtin/providers/aws/resource_aws_cloudfront_web_distribution.go

```diff
@@ -0,0 +1,587 @@
+			Prefix:         aws.String(d.Get("logging_prefix").(string)),
+			Bucket:         aws.String(d.Get("logging_bucket").(string)),
+		},
+		Restrictions: &cloudfront.Restrictions{
+			GeoRestriction: &cloudfront.GeoRestriction{
+				Quantity:        aws.Int64(int64(len(geoRestrictions))),
+				RestrictionType: aws.String(d.Get("geo_restriction_type").(string)),
+				Items:           geoRestrictions,
+			},
+		},
+		DefaultCacheBehavior: &cloudfront.DefaultCacheBehavior{
+			ForwardedValues: &cloudfront.ForwardedValues{
+				Cookies:     cookies,
+				QueryString: aws.Bool(d.Get("forward_query_string").(bool)),
+				Headers: &cloudfront.Headers{
+					Quantity: aws.Int64(int64(len(forwardedHeaders))),
+					Items:    forwardedHeaders,
+				},
+			},
+			TargetOriginId:       aws.String(originId),
+			ViewerProtocolPolicy: aws.String(d.Get("viewer_protocol_policy").(string)),
+			MinTTL:               aws.Int64(int64(d.Get("minimum_ttl").(int))),
+			MaxTTL:               aws.Int64(int64(d.Get("maximum_ttl").(int))),
+			DefaultTTL:           aws.Int64(int64(d.Get("default_ttl").(int))),
+			TrustedSigners: &cloudfront.TrustedSigners{
+				Enabled:  aws.Bool(false),
+				Quantity: aws.Int64(0),
+			},
+			SmoothStreaming: aws.Bool(d.Get("smooth_streaming").(bool)),
+			AllowedMethods: &cloudfront.AllowedMethods{
+				Quantity: aws.Int64(int64(len(allowedMethods))),
+				Items:    allowedMethods,
+				CachedMethods: &cloudfront.CachedMethods{
+					Quantity: aws.Int64(int64(len(cachedMethods))),
+					Items:    cachedMethods,
+				},
+			},
+		},
+		CacheBehaviors: &cloudfront.CacheBehaviors{
+			Quantity: aws.Int64(0),
+		},
+		CustomErrorResponses: &cloudfront.CustomErrorResponses{
+			Quantity: aws.Int64(0),
+		},
+	}, nil
+}
+
+// resourceAwsCloudFrontWebDistributionWaitUntilDeployed blocks until the
+// distribution is deployed. It currently takes exactly 15 minutes to deploy
+// but that might change in the future.
+func resourceAwsCloudFrontWebDistributionWaitUntilDeployed(
+	d *schema.ResourceData, meta interface{}) error {
+	stateConf := &resource.StateChangeConf{
+		Pending:    []string{"InProgress", "Deployed"},
+		Target:     "Deployed",
+		Refresh:    resourceAwsCloudFrontWebDistributionStateRefreshFunc(d, meta),
+		Timeout:    40 * time.Minute,
+		MinTimeout: 15 * time.Second,
+		Delay:      10 * time.Minute,
+	}
+
+	_, err := stateConf.WaitForState()
+	return err
+}
+
+func resourceAwsCloudFrontWebDistributionStateRefreshFunc(
+	d *schema.ResourceData, meta interface{}) resource.StateRefreshFunc {
+	return func() (interface{}, string, error) {
+		v, err := resourceAwsCloudFrontWebDistributionDistributionRetrieve(d, meta)
+
+		if err != nil {
+			log.Printf("Error on retrieving CloudFront distribution when waiting: %s", err)
+			return nil, "", err
+		}
+
+		if v == nil {
+			return nil, "", nil
+		}
+
+		return v.Distribution, *v.Distribution.Status, nil
+	}
+}
+
+func resourceAwsCloudFrontWebDistributionDistributionRetrieve(
```

Comment:
How about creating an extra `Computed` field `etag` for this and filling it in during `Read`? It may help us avoiding extra API calls and this whole function & extra logic in `Delete` & `Update`.

`Read` is typically called during `terraform refresh` which is by default called before each `terraform plan` or `terraform apply`, unless the user explicitly suppressed that, so I would expect the computed value from `Read` to be fairly up-to-date.